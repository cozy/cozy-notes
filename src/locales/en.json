{
  "Notes": {
    "Add": {
      "add_note": "Add a note"
    },
    "Delete": {
      "delete_note": "Delete",
      "deleted": "The note has been moved to the trash of your Drive",
      "failed": "The note could not be deleted, please try again"
    },
    "EditorView": {
      "main_placeholder": "Write anythingâ€¦"
    },
    "Editor": {
      "title_placeholder": "Here is your title",
      "exit_confirmation_message": "Automatic saving has not taken into account the latest changes in your doc. Please stay on your current note if you want to keep them.",
      "exit_confirmation_title": "Abandon your changes?",
      "exit_confirmation_leave": "Leave without saving",
      "exit_confirmation_cancel": "Cancel",
      "read_only": "This note is in read-only mode."
    },
    "Files": {
      "share": {
        "cta": "Share"
      }
    },
    "List": {
      "latest_notes": "Latest Notes",
      "name": "Name",
      "updated_at": "Updated at",
      "location": "Location",
      "sharings": "Sharings",
      "at": "%{date} at %{time}"
    },
    "Empty": {
      "welcome": "Welcome on Cozy Notes",
      "after_created": "Once your Note is created, you can find it in your "
    },
    "SavingIndicator": {
      "saved": {
        "default": "Up to date",
        "just_now": "Up to date",
        "secs_ago": "Up to date (saved %{time} sec. ago)",
        "min_ago": "Up to date (saved %{time} min. ago)",
        "mins_ago": "Up to date (saved %{time} min. ago)",
        "hour_ago": "Up to date (saved %{time} hour ago)",
        "hours_ago": "Up to date (saved %{time} hours ago)",
        "day_ago": "Up to date (saved %{time} day ago)",
        "days_ago": "Up to date (saved %{time} days ago)",
        "month_ago": "Up to date (saved %{time} month ago)",
        "monthes_ago": "Up to date (saved %{time} monthes ago)",
        "year_ago": "Up to date (saved %{time} year ago)",
        "years_ago": "Up to date (saved %{time} years ago)"
      },
      "saving": "Saving...",
      "still_saving": "Saving......",
      "out_of_sync": {
        "default": "Saved (off-line)",
        "just_now": "Saved (off-line)",
        "secs_ago": "Saved (off-line for %{time} sec)",
        "min_ago": "Saved (off-line for %{time} min)",
        "mins_ago": "Saved (off-line for %{time} min)",
        "hour_ago": "Saved (off-line for %{time} hour)",
        "hours_ago": "Saved (off-line for %{time} hours)",
        "day_ago": "Saved (off-line for %{time} day)",
        "days_ago": "Saved (off-line for %{time} days)",
        "month_ago": "Saved (off-line for %{time} month)",
        "monthes_ago": "Saved (off-line for %{time} monthes)",
        "year_ago": "Saved (off-line for %{time} year)",
        "years_ago": "Saved (off-line for %{time} years)"
      },
      "saving_impossible": {
        "default": "Changes unsaved (off-line)",
        "just_now": "Changes unsaved (off-line)",
        "secs_ago": "Changes unsaved (off-line for %{time} sec)",
        "min_ago": "Changes unsaved (off-line for %{time} min)",
        "mins_ago": "Changes unsaved (off-line for %{time} min)",
        "hour_ago": "Changes unsaved (off-line for %{time} hour)",
        "hours_ago": "Changes unsaved (off-line for %{time} hours)",
        "day_ago": "Changes unsaved (off-line for %{time} day)",
        "days_ago": "Changes unsaved (off-line for %{time} days)",
        "month_ago": "Changes unsaved (off-line for %{time} month)",
        "monthes_ago": "Changes unsaved (off-line for %{time} monthes)",
        "year_ago": "Changes unsaved (off-line for %{time} year)",
        "years_ago": "Changes unsaved (off-line for %{time} years)"
      }
    },
    "OfflineIndicator": {
      "message": "You're not connected to internet anymore. Your changes will be saved once you will be connected again."
    }
  },
  "Error": {
    "unshared_title": "This share is no longer active",
    "unshared_text": "The owner may have revoked this sharing. You can no longer edit a note at this address.",
    "loading_error_title": "The note couldn't be loaded",
    "loading_error_text_noReturnUrl": "The note may not exists anymore, or our servers may have a temporary glitch. Please try again later.",
    "loading_error_text_returnUrl": "The note may not exists anymore, or our servers may have a temporary glitch. Please go <a href='%{url}'>back</a> and try again later.",
    "422": "Only images are accepted",
    "unknown_error": "An unexpected error occured",
    "only_one_image": "You cannot drag and drop multiple pictures at the same time"
  },
  "manifest": {
    "short_description": "Cozy Notes is your personal and collaborative note-taking application.",
    "long_description": "Notes is an online text editor that can already offer you to:\n\n* Create your own notes with the essential features (bold, underlined, links, tables...)\n* Fill and find your notes with you other documents in files of your personal Drive\n* Edit easily your notes from all your devices\n* Access your notes offline\n* For the fans, use shortcuts and input in markdown format ... and even write with contacts together on the same note (coming soon)!",
    "screenshots": [
      "screenshots/en/screenshot01.png",
      "screenshots/en/screenshot02.png"
    ]
  }
}
