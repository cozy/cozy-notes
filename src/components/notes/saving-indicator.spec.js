import React from 'react'
import { I18n } from 'cozy-ui/transpiled/react/I18n'
import { mount } from 'enzyme'
import useBreakpoints from 'cozy-ui/transpiled/react/hooks/useBreakpoints'
import SavingIndicator from './saving-indicator'
import en from '../../locales/en.json'

jest.mock('cozy-ui/transpiled/react/hooks/useBreakpoints')

const sec = 1000
const min = 60 * sec
const hour = 60 * min
const day = 24 * hour

const getDirtySince = jest.fn()
const isDirty = () => !!getDirtySince()
const getLastSaveOrSync = jest.fn()
const on = jest.fn()

const collabProvider = {
  isDirty,
  getDirtySince,
  getLastSaveOrSync,
  on
}

global.document.createRange = jest.fn()

/**
 * Removes the varying part in MUI IDs,
 * which otherwise changes at each run
 *
 * See https://github.com/mui-org/material-ui/issues/21293
 *
 * @param {string} html
 * @returns {string}
 */
function normalizeMUI(html) {
  return html.replace(/(mui-[a-z0-9-]+)"/g, 'mui-"')
}

function itMatchSnapshot(collabProvider) {
  it('should match snapshot', () => {
    const component = mount(
      <SavingIndicator collabProvider={collabProvider} />,
      {
        wrappingComponent: I18n,
        wrappingComponentProps: {
          lang: 'en',
          dictRequire: () => en
        }
      }
    )
    expect(normalizeMUI(component.html())).toMatchSnapshot()
  })
}

describe('SavingIndicator', () => {
  afterEach(() => {
    jest.resetAllMocks()
  })
  beforeEach(() => {
    useBreakpoints.mockImplementation(() => ({
      isMobile: false
    }))

    // createRange is used by the MUI Tooltip for its placement
    global.document.createRange.mockImplementation(() => ({
      setStart: () => {},
      setEnd: () => {},
      commonAncestorContainer: {
        nodeName: 'BODY',
        ownerDocument: document
      }
    }))
  })

  describe('when dirtySince is truthy', () => {
    describe('when dirtySince is a few minutes ago', () => {
      beforeEach(() => {
        getDirtySince.mockImplementation(() => new Date(Date.now() - 15 * min))
      })
      itMatchSnapshot(collabProvider)
    })
    describe('when dirtySince is a few seconds ago', () => {
      beforeEach(() => {
        getDirtySince.mockImplementation(() => new Date(Date.now() - 2 * sec))
      })
      itMatchSnapshot(collabProvider)
    })
  })

  describe('when dirtySince is falsy', () => {
    const steps = {
      'few seconds': 2 * sec,
      '45 sec': 45 * sec,
      '1 min': 1 * min,
      '5 min': 5 * min,
      '15 min': 15 * min,
      '28 min': 28 * min,
      '50 min': 50 * min,
      '1 hour': 1 * hour,
      '3 hour': 3 * hour,
      '1 day': 1 * day,
      '10 day': 10 * day
    }
    Object.keys(steps).forEach(key => {
      describe(`when lastSave is ${key} ago`, () => {
        beforeEach(() => {
          getLastSaveOrSync.mockImplementation(
            () => new Date(Date.now() - steps[key] - 500)
          )
        })
        itMatchSnapshot(collabProvider)
      })
    })
  })

  describe('when offline', () => {
    beforeEach(() => {
      jest.spyOn(global.navigator, 'onLine', 'get').mockReturnValue(false)
    })
    // tooltip should be here
    itMatchSnapshot(collabProvider)
  })

  describe('when online', () => {
    beforeEach(() => {
      jest.spyOn(global.navigator, 'onLine', 'get').mockReturnValue(true)
    })
    // tooltip shouldn't be here
    itMatchSnapshot(collabProvider)
  })
})
