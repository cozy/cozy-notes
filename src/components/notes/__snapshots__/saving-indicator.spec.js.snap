// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavingIndicator when dirtySince is falsy when lastSave is 1 day ago should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\">Up to date</div>"`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 1 hour ago should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\">Up to date</div>"`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 1 min ago should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\">Up to date</div>"`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 3 hour ago should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\">Up to date</div>"`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 5 min ago should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\">Up to date</div>"`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 10 day ago should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\">Up to date</div>"`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 15 min ago should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\">Up to date</div>"`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 28 min ago should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\">Up to date</div>"`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 45 sec ago should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\">Up to date</div>"`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 50 min ago should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\">Up to date</div>"`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is few seconds ago should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\">Up to date</div>"`;

exports[`SavingIndicator when dirtySince is truthy when dirtySince is a few minutes ago should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\" aria-describedby=\\"mui-\\">Changes unsaved (off-line)</div><div role=\\"tooltip\\" class=\\"MuiTooltip-popper\\" id=\\"mui-\\" style=\\"position: absolute; top: 0px; left: 0px;\\" x-placement=\\"bottom\\"><div class=\\"MuiTooltip-tooltip MuiTooltip-tooltipPlacementBottom MuiTooltip-tooltipPlacementBottom\\" style=\\"opacity: 1; transform: scale(1, 1); transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 133ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\"><div class=\\"MuiGrid-root grid MuiGrid-container MuiGrid-wrap-xs-nowrap\\"><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"styles__c-chip___2aPqg styles__c-chip--normalSize___3U8N4 styles__c-chip--normalTheme___kVm5M u-breakword styles__c-chip--round___JCd2e chip\\"><svg class=\\"styles__icon___23x3R\\" style=\\"fill: var(--black);\\" width=\\"16\\" height=\\"16\\"><use xlink:href=\\"#offline\\"></use></svg></div></div><div class=\\"MuiGrid-root MuiGrid-item\\">You're not connected to internet anymore. Your changes will be saved once you will be connected again.</div></div></div></div>"`;

exports[`SavingIndicator when dirtySince is truthy when dirtySince is a few seconds ago should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\">Saving...</div>"`;

exports[`SavingIndicator when offline should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\" aria-describedby=\\"mui-\\">Saved (off-line)</div><div role=\\"tooltip\\" class=\\"MuiTooltip-popper\\" id=\\"mui-\\" style=\\"position: absolute; top: 0px; left: 0px;\\" x-placement=\\"bottom\\"><div class=\\"MuiTooltip-tooltip MuiTooltip-tooltipPlacementBottom MuiTooltip-tooltipPlacementBottom\\" style=\\"opacity: 1; transform: scale(1, 1); transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 133ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\"><div class=\\"MuiGrid-root grid MuiGrid-container MuiGrid-wrap-xs-nowrap\\"><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"styles__c-chip___2aPqg styles__c-chip--normalSize___3U8N4 styles__c-chip--normalTheme___kVm5M u-breakword styles__c-chip--round___JCd2e chip\\"><svg class=\\"styles__icon___23x3R\\" style=\\"fill: var(--black);\\" width=\\"16\\" height=\\"16\\"><use xlink:href=\\"#offline\\"></use></svg></div></div><div class=\\"MuiGrid-root MuiGrid-item\\">You're not connected to internet anymore. Your changes will be saved once you will be connected again.</div></div></div></div>"`;

exports[`SavingIndicator when online should match snapshot 1`] = `"<div class=\\"u-text saving-indicator\\">Up to date</div>"`;
